File: 1-anno/programmazione/lezioni/merge_sort.py
#! /usr/bin/python3


def merge_sort(vector,lx,rx):
    if lx < rx:
        mid = (lx+rx) // 2
        merge_sort(vector,lx,mid)
        merge_sort(vector,mid+1,rx)
        merge(vector,lx,mid,rx,key=helper)

def merge(vector,lx,mid,rx,key=None):
    i = lx
    j = mid + 1
    end = rx
    temp = []
    while i <= mid and j <= end:
        if key(vector[i]) < key(vector[j]):
            temp.append(vector[i])
            i += 1
        else:
            temp.append(vector[j])
            j += 1
    
    while i <= mid:
        temp.append(vector[i])
        i += 1

    while j <= end:
        temp.append(vector[j])
        j += 1
    
    i = lx

    #for e in temp:
    #    vector[i] = e
     #   i += 1

     # oppure utilizzando slicing

    vector[lx:lx+len(temp)] = temp[:]



def helper(item):
    if type(item) in (int,float):
        return (0,item)
    elif type(item) == str:
        return (1,item)



test = [3,4,'programmazione',8,99,0,-1,'ciao']
merge_sort(test,0,len(test)-1)
print(test)

File: 1-anno/programmazione/materiale/merge_sort.c
#include<stdio.h>
#include<stdlib.h>




void print_array(int* a,int lx, int n){
  printf("{ ");
  for(int i=lx; i < n; i++)
    printf("%d ",a[i]);

  printf(" }\n");

}

void merge(int* a,int s0,int l0,int s1, int l1){
  int size = (l1-s0)+1;
  int temp[size];
  int i = 0, tmp_s0 = s0;
  while(s0 <= l0 && s1 <= l1){
    if(a[s0] < a[s1]){
      temp[i] = a[s0];
      s0++;
    }else if(a[s0] > a[s1]){
      temp[i] = a[s1];
      s1++;
    }else{
      temp[i] = a[s0];
      s0++;
    }
    i++;
  }
  
  while(s0 <= l0){
    temp[i] = a[s0];
    s0++;
    i++;
  }
  while(s1 <= l1){
    temp[i] = a[s1];
    s1++;
    i++;
  }
  //print_array(temp,0,size);
  for(int i=0; i < size; i++){
    a[i+tmp_s0] = temp[i];
  }

}


void merge_sort(int *a,int lx,int n){
  if(n > lx){
    int mid = (lx+n)/2;
    merge_sort(a,lx,mid);
    merge_sort(a,mid+1,n);
    merge(a,lx,mid,mid+1,n);
  }
}





